### ISSUE TITLE
Bizarre duplication of metadata on different vods

### ISSUE THREAD

**Issue by EthanZeigler** [lay295/TwitchDownloader#1263]
### Checklist

- [X] I have checked the [issue page](https://github.com/lay295/TwitchDownloader/issues?q=is%3Aissue) for duplicates
- [X] I am running the latest version ([download here](https://github.com/lay295/TwitchDownloader/releases/latest))

### Edition

Command Line Interface

### Describe your issue here

I'm not sure how to describe this other than just showing.

I have been using this tool to download vods from a friend of mine for months and inserting everything into a searchable database. That isn't important for the bug but will make explaining the circumstances easier. I was using the latest version at the time of the bugs. The issues occurred in september and october.

Over this time period i've accumulated over 200 vod chats and while implementing file hashing to not reprocess data, I ran into this: 2 vod downloads where the metadata appears to have been merged and mangled in a way that makes absolutely no sense. For the record, these are **different vods**. They did not occur on the same day at the same time. They should be completely distinct. And yet, they're not?

I'm aware this is likely a bug on twitch's end, but its so unusual I felt i needed to put it here in case anyone else has run into the same behavior. The affected vods are in september and october of 2024. For privacy reasons I've redacted the name and ID of the streamer but can provide them privately if requested. Or you can just look up the vod id if it's floating around, but i wanted to avoid search engines indexing the name.

Does anyone know what on earth happened here???

```json
{
  "FileInfo": {
    "Version": {
      "Major": 1,
      "Minor": 3,
      "Patch": 1
    },
    "CreatedAt": "2024-09-16T23:21:16.9694971-04:00",
    "UpdatedAt": "0001-01-01T00:00:00"
  },
  "streamer": {
    "name": "<redacted>",
    "id": 00000
  },
  "video": {
    "title": "THE CRITTER IS: returning to my barista roots !LURK !DISCORD",
    "description": null,
    "id": "2252936436",
    "created_at": "2024-09-16T22:57:45Z",
    "start": 0,
    "end": 12966,
    "length": 12966,
    "viewCount": 20,
    "game": "The Closing Shift",
    "chapters": [
      {
        "id": "05f0636da035b42ca2411314878bde67",
        "startMilliseconds": 0,
        "lengthMilliseconds": 80000,
        "type": "GAME_CHANGE",
        "description": "The Closing Shift",
        "subDescription": "",
        "thumbnailUrl": "",
        "gameId": "1419063656",
        "gameDisplayName": "The Closing Shift",
        "gameBoxArtUrl": "https://static-cdn.jtvnw.net/ttv-boxart/1419063656_IGDB-40x53.jpg"
      },
      {
        "id": "1405752318707a2540edb9e88a1a54f5",
        "startMilliseconds": 80000,
        "lengthMilliseconds": 12886000,
        "type": "GAME_CHANGE",
        "description": "Art",
        "subDescription": "",
        "thumbnailUrl": "",
        "gameId": "509660",
        "gameDisplayName": "Art",
        "gameBoxArtUrl": "https://static-cdn.jtvnw.net/ttv-boxart/509660-40x53.jpg"
      }
    ]
  }
}
```

```json
{
  "FileInfo": {
    "Version": {
      "Major": 1,
      "Minor": 3,
      "Patch": 1
    },
    "CreatedAt": "2024-09-18T00:57:02.0421814-04:00",
    "UpdatedAt": "0001-01-01T00:00:00"
  },
  "streamer": {
    "name": "<redacted>",
    "id": 00000
  },
  "video": {
    "title": "THE CRITTER IS: chilllll drawinggg !LURK !DISCORD",
    "description": null,
    "id": "2252936436",
    "created_at": "2024-09-16T22:57:45Z",
    "start": 0,
    "end": 12966,
    "length": 12966,
    "viewCount": 1681,
    "game": "Art",
    "chapters": [
      {
        "id": "05f0636da035b42ca2411314878bde67",
        "startMilliseconds": 0,
        "lengthMilliseconds": 80000,
        "type": "GAME_CHANGE",
        "description": "Art",
        "subDescription": "",
        "thumbnailUrl": "",
        "gameId": "509660",
        "gameDisplayName": "Art",
        "gameBoxArtUrl": "https://static-cdn.jtvnw.net/ttv-boxart/509660-40x53.jpg"
      },
      {
        "id": "1405752318707a2540edb9e88a1a54f5",
        "startMilliseconds": 80000,
        "lengthMilliseconds": 12886000,
        "type": "GAME_CHANGE",
        "description": "Art",
        "subDescription": "",
        "thumbnailUrl": "",
        "gameId": "509660",
        "gameDisplayName": "Art",
        "gameBoxArtUrl": "https://static-cdn.jtvnw.net/ttv-boxart/509660-40x53.jpg"
      }
    ]
  }
}
```

**Comment by ScrubN** [comment]
Unfortunately because the VODs are expired I cannot do any debugging, however there are several details (such as the chapter ids) that suggest the issue was probably not created by TwitchDownloader.

**Comment by ScrubN** [comment]
For now, you will just need to manually fix it, however if you notice it again while the VOD is still available, let me know and I'll look into it.

**Comment by EthanZeigler** [comment]
Will do. Agreed, I dont think this is the downloader. It looks like twitch's API having a stroke in a completely nonsensical manner. The good news is the way ive set up my systems now, i'll be alerted as soon as a response like this is detected. If/when it happens, i'll update this issue asap.

**Comment by ScrubN** [comment]
It wouldn't be the first time Twitch's API delivered wrong or incomplete data. If this becomes a reoccurring issue, we will likely need to either change how we fetch VOD chapters or add logic to combine the extra chapters.

Checking TwitchTracker, it doesn't seem to have the same issues, but it might also use one of the other API calls for fetching VOD chapters.

**Comment by EthanZeigler** [comment]
It's not just the chapters. The vod ID is also nonsense.

**Comment by ScrubN** [comment]
Oh, I completely missed that. I genuinely have no clue how that could have happened as the ID stored is the exact same object as is used to fetch video info. The ID object is also never mutated, meaning this makes absolutely zero sense.

My only guess would have to be that at some point the file was mutated by another program.

**Comment by EthanZeigler** [comment]
That's why I'm suspicious of the twitch API. I'm using a modified version of https://github.com/cr08/TwitchVault to manage the data. It pulls a list of vods on the configured user and downloads them if they dont already exist. This seems to take into account the name associated to the vod as well, which technically could lead to an issue like this if the vod was to be renamed.

However, that's not what I'm seeing here because the data overlap is so bizarre. Clearly the API saw a vod with the same id as another at some point and downloaded the data associated with the id. That download had a different name and therefore wasn't seen automatically as a dupe by the wrapper. It invokes the twtichdownloader cli using the video ID as the source and sets the output to the vod id concatenated with the title as reported by twitch. In short, the twitch api did in fact at some point report a vod with this kind of bizarre data. The wrapper code and twitchdownloader hit the api independently and came to the same conclusion. I didn't link it here because it's not from this package, but the metadata reported by that wrapper agrees with what was in the chat download and the audio file it downloaded at the time is completely different.

TLDR i'm pretty confident twitch's api had a stroke. At least twice.

`ls -alh` output on the relevant files. The timestamp on `archive_chat.json` is the time the cli got invoked. The others can get a little weird because i'm lazy about the data processing and import sometimes.

```sh
-rw-r--r-- 1 ethan ethan 3.0M Oct  3 21:47 '20241002 T225528Z - 2266136455 - THE CRITTER IS HEARTGOLD ARTLOCKE LURK DISCORD_archive_chat.json'
-rw-r--r-- 1 ethan ethan  805 Dec  1 19:05 '20241002 T225528Z - 2266136455 - THE CRITTER IS HEARTGOLD ARTLOCKE LURK DISCORD_archive_info.json'
-rw-r--r-- 1 ethan ethan 3.2M Oct 26 03:39 '20241002 T225528Z - 2266136455 - THE CRITTER IS HEARTGOLD ARTLOCKE LURK DISCORD_archive_transcript.json'
-rw-r--r-- 1 ethan ethan 3.0M Oct  3 00:01 '20241002 T225528Z - 2266136455 - THE CRITTER IS MONSTER DESIGN CONTEST LURK DISCORD_archive_chat.json'
-rw-r--r-- 1 ethan ethan  778 Oct  3 00:00 '20241002 T225528Z - 2266136455 - THE CRITTER IS MONSTER DESIGN CONTEST LURK DISCORD_archive_info.json'
-rw-r--r-- 1 ethan ethan 3.2M Oct 26 03:43 '20241002 T225528Z - 2266136455 - THE CRITTER IS MONSTER DESIGN CONTEST LURK DISCORD_archive_transcript.json'
```
```sh
-rw-r--r-- 1 ethan ethan 3.9M Oct 23 22:16 '20241016 T225454Z - 2277847598 - THE CRITTER IS BACK TO NUZLOCKING GRIND LURK DISCORD_archive_chat.json'
-rw-r--r-- 1 ethan ethan  418 Dec  6 22:09 '20241016 T225454Z - 2277847598 - THE CRITTER IS BACK TO NUZLOCKING GRIND LURK DISCORD_archive_info.json'
-rw-r--r-- 1 ethan ethan 3.0M Oct 26 02:49 '20241016 T225454Z - 2277847598 - THE CRITTER IS BACK TO NUZLOCKING GRIND LURK DISCORD_archive_transcript.json'
-rw-r--r-- 1 ethan ethan 3.9M Oct 18 00:09 '20241016 T225454Z - 2277847598 - THE CRITTER IS ON THE COMMISSION GRIND LURK DISCORD_archive_chat.json'
-rw-r--r-- 1 ethan ethan  417 Oct 21 23:58 '20241016 T225454Z - 2277847598 - THE CRITTER IS ON THE COMMISSION GRIND LURK DISCORD_archive_info.json'
-rw-r--r-- 1 ethan ethan 3.0M Oct 26 03:04 '20241016 T225454Z - 2277847598 - THE CRITTER IS ON THE COMMISSION GRIND LURK DISCORD_archive_transcript.json'
```

**Comment by ScrubN** [comment]
Streamcharts encountered the same issue with the Twitch API. TwitchTracker correctly shows the they were 2 different streams; one on Sept. 13th and the other on Sept. 16th.

**Comment by EthanZeigler** [comment]
> After checking I think what may be happening.
> 
> The stream started at 16 September 2024 around 23:00.
> 5 minutes lates the game changed, but the title was also changed.
> 
> So what can be happening is the following:
> 
> Instead of having the metadata.txt like TwitchDownloaderCLI does, and choose only one of the titles for the main title, you split the metadata in two and converted to json, so there's one chapter on each one and with different title. But titles and games may not change at the same time. Both json are part of the same stream and have the same id.
> 
> The metadata.txt could include the game and the title in each chapter, but there's also the possibility that games and titles do not change at the same time so it will be more complex.
> 
> you can see the stream info here: https://streamscharts.com/channels/fourleafisland/streams/52006587565
> 
> the duration of each chapter is wrong in your json.
> 
> The metadata.txt once reconstructed will be something like:
> 
> ```
> ;FFMETADATA1
> title=THE CRITTER IS: returning to my barista roots !LURK !DISCORD (2252936436)
> artist=fourleafisland
> date=2024
> comment=Created at: 2024-09-16 22:57:45Z\
> Video id: 2252936436\
> Views: 1681
> [CHAPTER]
> TIMEBASE=1/1000
> START=0
> END=300000
> title=Game: The Closing Shift | Chapter: THE CRITTER IS: returning to my barista roots !LURK !DISCORD
> [CHAPTER]
> TIMEBASE=1/1000
> START=300000
> END=13200000
> title=Game: Art | Chapter: THE CRITTER IS: chilllll drawinggg !LURK !DISCORD
> ```
> 
> the game and title inside chapters have to be put in same place because there isn't other tags. Same for gameid and all the other if you want to add them but the title is too long.
> 
> the title change happens a lot when the guy starts a new stream but keeps the title from the previous one and updates it.
> 

Appreciate the effort! I think you've misunderstood the problem a bit though. This isn't a case of UTC rollover. That was one of the first things I checked. In fact, the api in question returns data in local time where the stream did not cross the date barrier. This is the twitch api returning bogus data for a stream id.

**Comment by EthanZeigler** [comment]
> Streamscharts encountered the same issue with the Twitch API. TwitchTracker correctly shows the they were 2 different streams; one on Sept. 13th and the other on Sept. 16th.

I guess my next question is why twitch tracker has this correct?

**Comment by superbonaci** [comment]
> Streamscharts encountered the same issue with the Twitch API. TwitchTracker correctly shows the they were 2 different streams; one on Sept. 13th and the other on Sept. 16th.

The stream from on Sept. 13th is not related to this issue, it's a different stream: https://twitchtracker.com/fourleafisland/streams/44810626811

The issue you are seeing is that streams with UTC rollover from 16th to 17th are displayed in both days in the table in twitchtracker, but once you enter the details it's placed in either of them not in both days.

For example twitchtracker:

![1](https://github.com/user-attachments/assets/72b45b65-5cfd-4eaa-bee0-7a7b9cbb2f01)
![2](https://github.com/user-attachments/assets/888b6517-e216-406b-a1a9-f435c1e26d9c)
![3](https://github.com/user-attachments/assets/96b93362-2b13-4538-83c8-0e1128bf9d3b)
![4](https://github.com/user-attachments/assets/f37fd4dc-0b34-45b7-a599-b0b6e450e372)

the stream will be placed on 17th sept. https://twitchtracker.com/fourleafisland/streams/52006587565
does not exist a stream placed on 16th with a id

Streamcharts places it on 16th, same id 52006587565: https://streamscharts.com/channels/fourleafisland/streams/52006587565

![5](https://github.com/user-attachments/assets/28642156-00d5-439f-859a-ce85077e32ae)

one places it the day when starts and other when ends but the only important time is the "Created at:"

**Comment by superbonaci** [comment]
The id 2252936436 is the same as 52006587565 because this one is for the vod and the other for the live stream. There's a Reddit post about it.

**Comment by superbonaci** [comment]
> Appreciate the effort! I think you've misunderstood the problem a bit though. This isn't a case of UTC rollover. That was one of the first things I checked. In fact, the api in question returns data in local time where the stream did not cross the date barrier. This is the twitch api returning bogus data for a stream id.

I don't understand exactly which is the bogus data, could you tell please?

**Comment by superbonaci** [comment]
I think you are not understanding it correctly, in both cases this is the same date and id:

```
 "id": "2252936436",
"created_at": "2024-09-16T22:57:45Z",
```

Because it's exactly the same video but there is one json for each chapter. The other date you see is when you created the json:

```
"CreatedAt": "2024-09-16T23:21:16.9694971-04:00",
"CreatedAt": "2024-09-18T00:57:02.0421814-04:00",
```

**Comment by superbonaci** [comment]
If you want to use trackers to do very accurate and in depth things forget about it, they are not as reliable as if you can catch the streams when are live on twitch or in the vods section.

They are just to double check in most cases it's ok.

**Comment by superbonaci** [comment]
The start and end times are both the same in both json, it's not calculating the corresponding times for each chapter:

```
    "start": 0,
    "end": 12966,
```

The program you are using to create those json does not work well. If it's [TwitchVault](https://github.com/cr08/TwitchVault) it hasn't been updated since Oct 19, 2023.

**Comment by EthanZeigler** [comment]
@superbonaci the program that's downloading those json files... is twitchdownloader. Twitchvault is a wrapper around twitchdownloader that automatically runs the download for unsaved vods, clips, chats, etc. Any other generation related to twitchvault is storing the raw information returned from the twitch graphql api directly in the same exact manner this project does.

Chapters are only handled within twitch downloader https://github.com/lay295/TwitchDownloader/blob/8bb2605d0630c6a4c5c1b9fb6c78c23d5423a9b8/TwitchDownloaderCore/TwitchHelper.cs#L1147

Twitch vault only stores basic metadata from the graphql api, however what information it does download corresponds with what twitch downloader stored as you can see in the examples above.
https://github.com/cr08/TwitchVault/blob/c5eda2d5e3d4bf6a6f0b9801f322916dfe143b05/utils.py#L84

**Comment by EthanZeigler** [comment]
TLDR this is on twitch and probably not something wither project can do anything about. I made this issue less for "downlaoder is messing up" and more for "be aware this is a thing twitch's api just decides to do"

**Comment by superbonaci** [comment]
But metadata.txt has the correct information. The only thing that is missing is when there are title change that only uses one of them not sure which one.

**Comment by ScrubN** [comment]
The chat json metadata files are prepared almost identically to the ffmpeg metadata files. I can assure you that this is not a problem with how TD handles data from Twitch.

**Comment by EthanZeigler** [comment]
I have another example of bizarre behavior now from november 30th, but i need to double check what we're seeing is the same issue before sharing

**Comment by superbonaci** [comment]
Yes if you are going to share it make sure it's a vod that we can download and check.

**Comment by EthanZeigler** [comment]
```json
{
    "id": "2340216348",
    "user_id": "436970891",
    "user_name": "<redacted>",
    "title": "doodlin! !DONO !LURK !DISCORD",
    "type": "archive",
    "duration": "2h32m27s",
    "url": "https://www.twitch.tv/videos/2340216348",
    "views": 6,
    "moments": [
        {
            "duration": 84,
            "offset": 0,
            "id": "509660",
            "name": "Art",
            "type": "GAME_CHANGE"
        },
        {
            "duration": 0,
            "offset": 84,
            "id": "2085980140",
            "name": "Lethal Company",
            "type": "GAME_CHANGE"
        }
    ],
    "muted_segments": [],
    "recorded_at": "2024-12-31T00:02:32Z",
    "recorded_at_iso": "20241231 T000232Z"
}
```

```json
{
    "id": "2340216348",
    "user_id": "436970891",
    "user_name": "<redacted>",
    "title": "LETHAL COMPANY CHRISTMAS !DONO !LURK !DISCORD",
    "type": "archive",
    "duration": "2h32m27s",
    "url": "https://www.twitch.tv/videos/2340216348",
    "views": 1572,
    "moments": [
        {
            "duration": 84,
            "offset": 0,
            "id": "2085980140",
            "name": "Lethal Company",
            "type": "GAME_CHANGE"
        },
        {
            "duration": 9063,
            "offset": 84,
            "id": "2085980140",
            "name": "Lethal Company",
            "type": "GAME_CHANGE"
        }
    ],
    "muted_segments": [],
    "recorded_at": "2024-12-31T00:02:32Z",
    "recorded_at_iso": "20241231 T000232Z"
}
```

![image](https://github.com/user-attachments/assets/06937b8c-097b-4c5b-b6ce-9e0c0a9644a9)
The `doodlin` vod (pokemon from memory) got overwritten with the lethal company one.

**Comment by superbonaci** [comment]
This is the metadata.txt:

```
;FFMETADATA1
title=LETHAL COMPANY CHRISTMAS !DONO !LURK !DISCORD (2340216348)
artist=redacted
date=2024
genre=Lethal Company
comment=Created at: 2024-12-31 00:02:32Z\
Video id: 2340216348\
Views: 1588
[CHAPTER]
TIMEBASE=1/1000
START=0
END=84000
title=Lethal Company
[CHAPTER]
TIMEBASE=1/1000
START=84000
END=9147000
title=Lethal Company
```

**Comment by superbonaci** [comment]
I don't know if there can be any bug when title and game don't change at the same time, for example:

0 - start with one title and one game
1h - change title
2h - change game
3h - end stream

The metadata.txt will only have one title and the 2 games (the chapters), but if you try to to be very precise and log everything some bugs may happen. I don't know if the Twitch API can be that accurate, all that information should be available somehow after the stream ends, mabe @ScrubN knows...

@EthanZeigler What is that you want to do exactly, what do you find wrong?

**Comment by superbonaci** [comment]
It's working fine now, I think you just need to wait until the stream ends and Twitch processes correctly the video. Also if there are any after end edits it should be as well.

```
{
  "data": {
    "video": {
      "viewCount": 1591,
      "moments": {
        "pageInfo": {
          "hasNextPage": false
        },
        "edges": [
          {
            "node": {
              "details": {
                "game": {
                  "id": "2085980140",
                  "displayName": "Lethal Company",
                  "name": "Lethal Company"
                }
              },
              "positionMilliseconds": 0,
              "durationMilliseconds": 84000,
              "type": "GAME_CHANGE"
            }
          },
          {
            "node": {
              "details": {
                "game": {
                  "id": "2085980140",
                  "displayName": "Lethal Company",
                  "name": "Lethal Company"
                }
              },
              "positionMilliseconds": 84000,
              "durationMilliseconds": 9063000,
              "type": "GAME_CHANGE"
            }
          }
        ]
      }
    }
  },
  "extensions": {
    "durationMilliseconds": 29,
    "requestID": "<redacted>"
  }
}
```

**Comment by EthanZeigler** [comment]
interesting. that gives me a theory on what might be happening then. Is it possible the automation for finding the channel ids is running into a race condition with the vod finalization process? Will experiment and report back in the future.

**Comment by superbonaci** [comment]
You mean the game ID, which is always the same for each game? The channel and the user have their own ID but that's a different thing.

```
                "game": {
                  "id": "2085980140",
                  "displayName": "Lethal Company",
                  "name": "Lethal Company"
                }
```

```
        {
            "duration": 84,
            "offset": 0,
            "id": "509660",
            "name": "Art",
            "type": "GAME_CHANGE"
        },
```

**Comment by superbonaci** [comment]
I think the duration 0:

```
    "moments": [
        {
            "duration": 84,
            "offset": 0,
            "id": "509660",
            "name": "Art",
            "type": "GAME_CHANGE"
        },
        {
            "duration": 0,
            "offset": 84,
            "id": "2085980140",
            "name": "Lethal Company",
            "type": "GAME_CHANGE"
        }
    ],
```

Is because you are reading the info while the stream is live, check here:
https://github.com/lay295/TwitchDownloader/blob/b51f7245a9bf38ddccbe6b40a35347e7b27d9aa8/TwitchDownloaderCore/TwitchHelper.cs#L1142

**Comment by EthanZeigler** [comment]
Okay, i've confirmed it. It's a race condition bug within the twitch api.

Between when a stream has ended and the vod is still being finalized, the api returns incorrect information for the _previous_ vod. This leads to the script i'm using identifying the already downloaded vod as different one than previously downloaded since the title and first chapter changes. It redownloads that vod data and stores it under the incorrect name as returned by the api.

For example, if the just finished stream is going to be assigned id 345 and the last recorded and available vod is `123`, calls to the twitch api for `123` returns `123`'s data with the first chapter information of vod `345` overwriting `123`'s and the title of `345` being overwritten by `345`'s first chapter title. After the vod is finalized (and presumably if there's a cache it has flushed), `123` begins returning correct data once again. 

In a more distilled form,
vod  123, titled "abc", chapter 1 titled "uvw"
vod 345, titled "cde", chapter 1 titled "xyz"

## Stream running
```py
vods = api.get_vods()
-> [123]

for vod in vods:
    if vod <timestamp>_<id>_<title> exists:
        print("vod {vod} exists")
        continue
    else:
        print("vod {vod} downloading")
        api.download_vod(vod)
-> "vod 123|abc exists"
```

## Stream ended, vod finalized
```py
vods = api.get_vods()
-> [123, 345]

for vod in vods:
    if vod <timestamp>_<id>_<title> exists:
        print("vod {vod} exists")
        continue
    else:
        print("vod {vod} downloading")
        api.download_vod(vod)
-> "vod 123|abc exists\nvod 345|cde downloading"
```

## Stream ended, vod not finalized (bug)
```py
vods = api.get_vods()
-> [123]

for vod in vods:
    if vod <timestamp>_<id>_<title> exists:
        print("vod {vod} exists")
        continue
    else:
        print("vod {vod} downloading")
        api.download_vod(vod)
-> "vod 123|xyz downloading"
```

**Comment by EthanZeigler** [comment]
No. It's affected by the same upstream bug.

**Comment by superbonaci** [comment]
If you could give examples with real ids, I don't understand exactly what you mean. The "last recorded and available vod" is another vod from the same guy, or is the id when the stream is ongoing? for example 43477083080 and 2340216348?
because when the stream ends the older id is no longer available.

or you mean it mixes 2341101563 and 2340216348?

for what i could see, the vods are finalized immediately, it just updates the playlist to include the latest part and creates a thumbnails instead of the question mark '?' but it happens at the very moment the live stream ends.

maybe when the stream ends it has to choose one of the titles of the chapters to be the title of the vod, or maybe it was the las one used, I think you've got confused somehow.

If you could explain with a real example it would be much easier for us to see, instead of being so cryptic about it.
Pick any stream that you don't care about, I don't see what could be the problem.
